// pulseflow-server/controllers/authController.js
import User from '../models/User.js';
import bcrypt from 'bcrypt';
import otpService from '../services/otpService.js';
import tokenService from '../services/tokenService.js';
import jwt from 'jsonwebtoken';
import { sendVerificationCode } from '../services/emailService.js';

export const register = async (req, res) => {
  try {
    const { senha, email } = req.body;

    const userExists = await User.findOne({ email });
    if (userExists) {
      return res.status(400).json({ message: 'Usuário já existe.' });
    }

    const hashedPassword = await bcrypt.hash(senha, 10);
    const otp = otpService.generateOTP();

    const newUser = new User({
      ...req.body,
      senha: hashedPassword,
      otp: otp.code,
      otpExpires: otp.expires,
    });

    await newUser.save();
    res.status(201).json({ message: 'Usuário registrado com sucesso!' });
  } catch (err) {
    res.status(500).json({ message: 'Erro ao registrar.', error: err.message });
  }
};

export const login = async (req, res) => {
  try {
    const { email, senha } = req.body;
    const user = await User.findOne({ email });

    if (!user) return res.status(401).json({ message: 'Usuário não encontrado.' });
    const isMatch = await bcrypt.compare(senha, user.senha);
    if (!isMatch) return res.status(401).json({ message: 'Senha incorreta.' });

    const token = tokenService.generateToken({ id: user._id, email: user.email });
    res.status(200).json({ message: 'Login realizado com sucesso!', token });
  } catch (err) {
    res.status(500).json({ message: 'Erro ao fazer login.', error: err.message });
  }
};

export const sendCode = async (req, res) => {
  try {
    const { method } = req.body;
    const token = req.headers.authorization?.split(' ')[1];
    const decoded = jwt.verify(token, process.env.JWT_SECRET);
    const user = await User.findById(decoded.id);

    if (!user) return res.status(404).json({ message: 'Usuário não encontrado.' });

    const otp = otpService.generateOTP();
    user.otp = otp.code;
    user.otpExpires = otp.expires;
    await user.save();

    if (method === 'email') {
      await sendVerificationCode(user.email, otp.code);
    }

    res.status(200).json({ message: `Código enviado via ${method}.` });
  } catch (err) {
    res.status(500).json({ message: 'Erro ao enviar código.', error: err.message });
  }
};

export const verifyOTP = async (req, res) => {
  try {
    const { method, code } = req.body;
    const token = req.headers.authorization?.split(' ')[1];
    const decoded = jwt.verify(token, process.env.JWT_SECRET);
    const user = await User.findById(decoded.id);

    if (!user || !user.otp || !user.otpExpires) {
      return res.status(400).json({ message: 'Código não encontrado.' });
    }

    if (user.otp !== code) {
      return res.status(400).json({ message: 'Código incorreto.' });
    }

    if (user.otpExpires < new Date()) {
      return res.status(400).json({ message: 'Código expirado.' });
    }

    user.otp = null;
    user.otpExpires = null;
    await user.save();

    res.status(200).json({ message: 'Verificação bem-sucedida!' });
  } catch (err) {
    res.status(500).json({ message: 'Erro ao verificar código.', error: err.message });
  }
};
